File: persistence/src/main/java/de/cyface/persistence/AbstractCyfaceTable.java
package de.cyface.persistence;

import android.content.ContentValues;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;

import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Abstract base class for all Cyface measurement tables implementing common functionality.
 *
 * @author Klemens Muthmann
 * @author Armin Schnabel
 * @version 1.1.0
 * @since 1.0.0
 */
public abstract class AbstractCyfaceTable implements CyfaceTable {

    /**
     * The database table name.
     */
    private final String name;

    AbstractCyfaceTable(final String name) {
        if (name.isEmpty()) {
            throw new IllegalStateException("Database table name may not be empty.");
        }

        this.name = name;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public final long insertRow(final SQLiteDatabase database, final ContentValues values) {
        return database.insert(getName(), null, values);
    }

    // BulkInsert is about 80 times faster than insertBatch
    @Override
    public final long[] insertBatch(SQLiteDatabase db, final List<ContentValues> valuesList) {
        long[] ret = new long[valuesList.size()];
        db.beginTransaction();
        try {
            int len = valuesList.size();
            for (int i = 0; i < len; i++) {
                ret[i] = insertRow(db, valuesList.get(i));
            }
            db.setTransactionSuccessful();
        } finally {
            db.endTransaction();
        }
        return ret;
    }

    @Override
    public final int deleteRow(final SQLiteDatabase database, final String selection, final String[] selectionArgs) {
        return database.delete(getName(), selection, selectionArgs);
    }

    @Override
    public void onCreate(final SQLiteDatabase database) {
        database.execSQL(getCreateStatement());
    }

    protected abstract String getCreateStatement();

    @Override
    public abstract void onUpgrade(final SQLiteDatabase database, final int oldVersion, final int newVersion);

    @Override
    public Cursor query(final SQLiteDatabase database, final String[] projection, final String selection,
                        final String[] selectionArgs, final String sortOrder) {
        checkColumns(projection);
        /*LOGGER.debug("Querying database table {} with projection {} selection {} and arguments {} limit {} isACountingQuery: {}",
                getName(), projection, selection, Arrays.toString(selectionArgs), queryLimit, isACountingQuery(projection));*/
        return database.query(getName(), projection, selection, selectionArgs, null, null, sortOrder);
    }

    protected void checkColumns(String[] projection) {
        if (projection != null) {
            Set<String> requestedColumns = new HashSet<>(Arrays.asList(projection));
            Set<String> availableColumns = new HashSet<>(Arrays.asList(getDatabaseTableColumns()));
            if (!availableColumns.containsAll(requestedColumns)) {
                throw new IllegalArgumentException("Unknown columns in projection");
            }
        }
    }

    protected abstract String[] getDatabaseTableColumns();

    @Override
    public int update(SQLiteDatabase database, ContentValues values, String selection,
                      String[] selectionArgs) {
        return database.update(getName(), values, selection, selectionArgs);
    }

    @Override
    public String toString() {
        return name;
    }
}

File: persistence/src/main/java/de/cyface/persistence/content/AbstractCyfaceTable.kt
/*
 * Copyright 2017-2023 Cyface GmbH
 *
 * This file is part of the Cyface SDK for Android.
 *
 * The Cyface SDK for Android is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * The Cyface SDK for Android is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with the Cyface SDK for Android. If not, see <http://www.gnu.org/licenses/>.
 */
package de.cyface.persistence.content

import android.content.ContentValues
import android.database.Cursor
import android.database.sqlite.SQLiteDatabase.CONFLICT_NONE
import android.database.sqlite.SQLiteQueryBuilder
import androidx.sqlite.db.SupportSQLiteDatabase


/**
 * Abstract base class for all Cyface measurement tables implementing common functionality.
 *
 * @author Klemens Muthmann
 * @author Armin Schnabel
 * @version 2.0.0
 * @since 1.0.0
 */
abstract class AbstractCyfaceTable internal constructor(name: String) : CyfaceTable {
    /**
     * The database table name.
     */
    final override val name: String

    /**
     * An array containing all the column names used by this table.
     */
    protected abstract val databaseTableColumns: Array<String>

    init {
        check(name.isNotEmpty()) { "Database table name may not be empty." }
        this.name = name
    }

    override fun insertRow(database: SupportSQLiteDatabase, values: ContentValues): Long {
        // CONFLICT_NONE is used as this was the default behavior before (in SQLiteDatabase)
        return database.insert(name, CONFLICT_NONE, values)
        // We could use Room, but stick with the tested low-level implementation for now.
        // Room implementation: https://github.com/android/architecture-components-samples/
    }

    // BulkInsert is about 80 times faster than insertBatch
    override fun insertBatch(
        database: SupportSQLiteDatabase,
        valuesList: List<ContentValues>
    ): LongArray {
        val ret = LongArray(valuesList.size)
        database.beginTransaction()
        try {
            val len = valuesList.size
            for (i in 0 until len) {
                ret[i] = insertRow(database, valuesList[i])
            }
            database.setTransactionSuccessful()
        } finally {
            database.endTransaction()
        }
        return ret
    }

    override fun deleteRow(
        database: SupportSQLiteDatabase,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        return database.delete(name, selection, selectionArgs)
    }

    override fun query(
        database: SupportSQLiteDatabase, projection: Array<String>?, selection: String?,
        selectionArgs: Array<String>?, sortOrder: String?
    ): Cursor {
        checkColumns(projection)
        // We could use a Room generated cursor: https://stackoverflow.com/a/46883664/5815054
        val builder = SQLiteQueryBuilder()
        builder.tables = name
        val query = builder.buildQuery(projection, selection, null, null, sortOrder, null)
        return database.query(query)
    }

    private fun checkColumns(projection: Array<String>?) {
        if (projection != null) {
            val requestedColumns: Set<String> = HashSet(listOf(*projection))
            val availableColumns: Set<String> = HashSet(mutableListOf(*databaseTableColumns))
            require(availableColumns.containsAll(requestedColumns)) { "Unknown columns in projection" }
        }
    }

    override fun update(
        database: SupportSQLiteDatabase,
        values: ContentValues?,
        selection: String?,
        selectionArgs: Array<String>?
    ): Int {
        // CONFLICT_NONE is used as this was the default behavior before (in SQLiteDatabase)
        return database.update(name, CONFLICT_NONE, values, selection, selectionArgs)
    }

    override fun toString(): String {
        return name
    }

    companion object {
        /**
         * Loading all entries at once seems slower than loading it in chunks of 10k entries [#MOV-248]
         */
        const val DATABASE_QUERY_LIMIT = 10000
    }
}